lifecycle-coroutines/src/commonMain/kotlin/com/arkivanov/essenty/lifecycle/coroutines/LifecycleCoroutinesExt.kt

```diff
@@ -0,0 +1,147 @@
+            suspendCancellableCoroutine { cont ->
+                callback = createLifecycleAwareCallback(
+                    startState = state,
+                    onStateAppear = {
+                        job = startJob(mutex, block)
+                    },
+                    onStateDisappear = {
+                        job?.cancel()
+                        job = null
+                    },
+                    onDestroy = { cont.resume(Unit) },
+                )
+
+                this@repeatOnEssentyLifecycle.subscribe(callback as Lifecycle.Callbacks)
+            }
+        } finally {
+            job?.cancel()
+            job = null
+            callback?.let {
+                this@repeatOnEssentyLifecycle.unsubscribe(it)
+            }
+        }
+    }
+}
+
+/**
+ * Start [Flow] collecting when [Lifecycle.State] is at least as [minActiveState].
+ * It stopped collecting if "opposite" [Lifecycle.State] will appear.
+ */
+fun <T> Flow<T>.flowWithEssentyLifecycle(
+    lifecycle: Lifecycle,
+    minActiveState: Lifecycle.State = Lifecycle.State.STARTED
+): Flow<T> = callbackFlow {
+    lifecycle.repeatOnEssentyLifecycle(minActiveState) {
+        this@flowWithEssentyLifecycle.collect {
+            send(it)
+        }
+    }
+    close()
+}
+
+/**
+ * Function for executing [block] on the [CoroutineScope] with passed specific [dispatcher].
+ */
+private suspend fun withCoroutineScope(
+    dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
+    block: suspend CoroutineScope.() -> Unit
+): Unit = coroutineScope {
+    withContext(dispatcher) {
+        block()
+    }
+}
+
+/**
+ * Creates lifecycle aware [Lifecycle.Callbacks] interface instance.
+ *
+ * @param startState [Lifecycle.State] that [onStateAppear] block must be called from
+ * @param onStateAppear block of code that will be executed when the [Lifecycle.State] was equal [startState]
+ * @param onStateDisappear block of code that will be executed when the [Lifecycle.State] was equal to opposite [startState]
+ * @param onDestroy block of code that will be executed when the [Lifecycle.State] was equal [Lifecycle.State.DESTROYED]
+ *
+ * @return [Lifecycle.Callbacks]
+ */
+private fun createLifecycleAwareCallback(
+    startState: Lifecycle.State,
+    onStateAppear: () -> Unit,
+    onStateDisappear: () -> Unit,
+    onDestroy: () -> Unit,
+): Lifecycle.Callbacks = object : Lifecycle.Callbacks {
+
+    override fun onCreate(): Unit = launchWithState(Lifecycle.State.CREATED)
+
+    override fun onStart(): Unit = launchWithState(Lifecycle.State.STARTED)
+
+    override fun onResume(): Unit = launchWithState(Lifecycle.State.RESUMED)
+
+    override fun onPause(): Unit = closeWithState(Lifecycle.State.RESUMED)
+
+    override fun onStop(): Unit = closeWithState(Lifecycle.State.STARTED)
+
+    override fun onDestroy() {
+        closeWithState(Lifecycle.State.CREATED)
+        onDestroy()
+    }
+
+    private fun launchWithState(launchState: Lifecycle.State) {
```

Comment:
It looks like a better name would be `launchIfState`, because it basically checks `if (startState == launchState)`. And the argument name could be just `state` in this case. Same for `closeWithState` -> `closeIfState`.

Reply:
Corrected.